design/ceph/node-loss-rbd-rwo.md

```diff
@@ -0,0 +1,123 @@
+
+## Proposed Solution
+
+> Note: This solution requires minimum kubernetes version 1.26.0
+
+The kubernetes feature [Non-Graceful Node Shutdown](https://kubernetes.io/blog/2022/12/16/kubernetes-1-26-non-graceful-node-shutdown-beta/) is available starting in K8s 1.26 to help improve the volume recovery during node loss. When a node is lost, the admin is required to add the taint `out-of-service` manually to the node. After the node is tainted, K8s will:
+
+- Remove the volume attachment from the lost node
+- Delete the old pod on the lost node
+- Create a new pod on the new node
+- Allow the volume to be attached to the new node
+
+Once this taint is applied manually, rook will create a [NetworkFence CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md). The [csi-addons operator](https://github.com/csi-addons/kubernetes-csi-addons) will then blocklist the node to prevent any ceph rbd client on the lost node from writing any more data.
+
+After the new pod is running on the new node and the old node which was lost comes back, rook will delete the [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md).
+
+example of taint to be applied to lost node:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule
+```
+
+> Note: This will be enabled by default in rook if the NetworkFence CR is found, in the case for some reason user wants to disable this feature in rook can edit `rook-ceph-operator-config` configmap and update the `ROOK_WATCH_FOR_NODE_FAILURE: "false"`.
+
+## How get which CIDR to block
+
+There are multiple networking options available for example, host Networking, Pod networking, Multus etc. This make it difficult to know which NodeIp address to blockList.
+For this we'll follow the following approach which will work for all networking options(but still will not work for external cluster which nodeIps):
+
+1. Get the `volumesInUse` from the node which has the taint `out-of-service`.
+
+2. List all the pv and compare the pv `spec.volumeHandle` with the node `volumesInUse` field `volumeHandle`
+   1. Example:
+
+      Below is sample Node volumeInUse field
+
+    ```
+      volumesInUse:
+        - kubernetes.io/csi/rook-ceph.rbd.csi.ceph.com^0001-0009-rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002
+
+      # Note: The volumeInUse naming convention are `kubernetes.io/csi/ + CSI driver name + ^ + volumeHandle`
+    ```
+
+     and the following is pv `volumeInHandle`
+
+     ```
+      volumeHandle: 0001-0009-rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002
+    ```
```

Comment:
```suggestion
   1. Example:

      Below is sample Node volumeInUse field

    ```
      volumesInUse:
        - kubernetes.io/csi/rook-ceph.rbd.csi.ceph.com^0001-0009-rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002

      # Note: The volumeInUse naming convention are `kubernetes.io/csi/ + CSI driver name + ^ + volumeHandle`
    ```

     and the following is pv `volumeInHandle`

     ```
      volumeHandle: 0001-0009-rook-ceph-0000000000000002-24862838-240d-4215-9183-abfc0e9e4002
    ```
```

@travisn @Rakshith-R I added this section to clear `volumeInUse` and `volumeHandle` names.